Add support for assign unique id node and filter in our join stats

sopel39 opened the pull request:
Q21 before:
```
- LeftJoin[("orderkey" = "orderkey_10") AND ("suppkey_12" <> "suppkey_0")][$hashvalue_204, $hashv
        Distribution: REPLICATED
        Cost: {rows: 2450757425 (228.24GB), cpu: 2391276411339.29, memory: 255027000671.95, netwo
    - Project[] => [orderkey:bigint, suppkey_0:bigint, commitdate:date, receiptdate:date, name:va
            Cost: {rows: 296052124 (27.57GB), cpu: 1276596897180.48, memory: 15027412311.95, netw
            $hashvalue_204 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey
        - AssignUniqueId => [orderkey:bigint, suppkey_0:bigint, commitdate:date, receiptdate:date
                Cost: {rows: 296052124 (24.81GB), cpu: 1246991684800.55, memory: 15027412311.95,
            - Project[] => [orderkey:bigint, suppkey_0:bigint, commitdate:date, receiptdate:date,
                    Cost: {rows: 296052124 (22.06GB), cpu: 1244623267810.15, memory: 15027412311.
                - InnerJoin[("orderkey" = "orderkey_3")][$hashvalue_201, $hashvalue_202] => [orde
                        Distribution: REPLICATED
                        Cost: {rows: 296052124 (24.81GB), cpu: 1220939097906.21, memory: 15027412
                        ...
    - LocalExchange[HASH][$hashvalue_205] ("orderkey_10") => suppkey_12:bigint, orderkey_10:bigin
            Cost: {rows: 5999989709 (223.52GB), cpu: 599998970900.00, memory: 0.00, network: 0.00
        - ScanProject[table = hive:tpch_sf1000_orc:lineitem, originalConstraint = true] => [suppk
                Cost: {rows: 5999989709 (111.76GB), cpu: 119999794180.00, memory: 0.00, network:
                non_null_188 := true
                $hashvalue_206 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orde
                LAYOUT: tpch_sf1000_orc.lineitem
                suppkey_12 := HiveColumnHandle{name=suppkey, hiveType=bigint, hiveColumnIndex=2,
```
after:
```
- RightJoin[("orderkey_10" = "orderkey") AND ("suppkey_12" <> "suppkey_0")][$hashvalue_195, $hash
        Distribution: PARTITIONED
        Cost: {rows: 2450757425 (228.24GB), cpu: 2181375456974.36, memory: 44879335499.45, networ
    - ScanProject[table = hive:tpch_sf1000_orc:lineitem, originalConstraint = true] => [suppkey_1
            Cost: {rows: 5999989709 (111.76GB), cpu: 119999794180.00, memory: 0.00, network: 0.00
            non_null_188 := true
            $hashvalue_195 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orderkey
            LAYOUT: tpch_sf1000_orc.lineitem
            suppkey_12 := HiveColumnHandle{name=suppkey, hiveType=bigint, hiveColumnIndex=2, colu
            orderkey_10 := HiveColumnHandle{name=orderkey, hiveType=bigint, hiveColumnIndex=0, co
    - LocalExchange[HASH][$hashvalue_196] ("orderkey") => orderkey:bigint, suppkey_0:bigint, comm
            Cost: {rows: 296052124 (27.57GB), cpu: 1306695531175.55, memory: 15274123119.52, netw
        - Project[] => [orderkey:bigint, suppkey_0:bigint, commitdate:date, receiptdate:date, nam
                Cost: {rows: 296052124 (27.57GB), cpu: 1277090318795.61, memory: 15274123119.52,
                $hashvalue_206 := "combine_hash"(BIGINT '0', COALESCE("$operator$hash_code"("orde
            - AssignUniqueId => [orderkey:bigint, suppkey_0:bigint, commitdate:date, receiptdate:
                    Cost: {rows: 296052124 (24.81GB), cpu: 1247485106415.68, memory: 15274123119.
                - Project[] => [orderkey:bigint, suppkey_0:bigint, commitdate:date, receiptdate:d
                        Cost: {rows: 296052124 (22.06GB), cpu: 1245116689425.29, memory: 15274123
                    - InnerJoin[("orderkey" = "orderkey_3")][$hashvalue_203, $hashvalue_204] => [
                            Distribution: PARTITIONED
                            Cost: {rows: 296052124 (24.81GB), cpu: 1221432519521.34, memory: 1527
                        - Project[] => [orderkey:bigint, suppkey_0:bigint, commitdate:date, recei
                                Cost: {rows: 296052124 (22.06GB), cpu: 1066103658475.45, memory:
                                $hashvalue_203 := "combine_hash"(BIGINT '0', COALESCE("$operator$
```

Basically outer join sides were not flipped.

rschlussel2 commented on presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java:
```diff
@@ -241,15 +242,10 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
         if (criteria.isEmpty()) {
             if (rightStats.getOutputRowCount() > 0) {
                 // all left side rows are matched
+                // TODO: account non-equi conditions here somehow
```
maybe using the UNKNOWN_FILTER_COEFFICIENT, same as if there's a mix of equi and non-equi conditions.

rschlussel2 commented on presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java:
```diff
@@ -318,10 +318,8 @@ else if (leftNDV <= matchingRightNDV) {
-            result = result.mapOutputRowCount(rowCount -> min(leftStats.getOutputRowCount(), rowCount / UNKNOWN_FILTER_COEFFICIENT));
-        }
+        // limit the number of complement rows (to left row count) and account for remaining clauses
+        result = result.mapOutputRowCount(rowCount -> min(leftStats.getOutputRowCount(), rowCount / Math.pow(UNKNOWN_FILTER_COEFFICIENT, numberOfRemainingClauses)));
```
This also affects equijoins with 3+ criteria.  Did you notice any effect on those kinds of queries?

findepi replied:
@rschlussel2  You mean that for 3 criteria the computed `result` will be different?

rschlussel2 replied:
yes

sopel39 replied:
> This also affects equijoins with 3+ criteria. Did you notice any effect on those kinds of queries?

The code treats equicriteria the same as it was before this change. I just unified how non-equi criteria are treated.

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java:
```diff
@@ -119,6 +121,12 @@ public PlanNodeCostEstimate visitGroupReference(GroupReference node, Void contex
+        @Override
+        public PlanNodeCostEstimate visitAssignUniqueId(AssignUniqueId node, Void context)
+        {
+            return cpuCost(getStats(node.getSource()).getOutputRowCount() * BigintType.BIGINT.getFixedSize());
```
nit: `getStats(node)` should be sufficient (instead of `getStats(node.getSource())`) -- provided that we have stats rule for `AssignUniqueId`, which we need to have anyway.

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/AssignUniqueIdStatsRule.java:
```diff
@@ -0,0 +1,50 @@
+                .addSymbolStatistics(assignUniqueId.getIdColumn(), SymbolStatsEstimate.builder()
+                        .setDistinctValuesCount(sourceStats.getOutputRowCount())
+                        .setNullsFraction(0.0)
+                        .build())
```
- `setAverageRowSize` ?
- Can we say eg that the `assignUniqueId.getIdColumn()` symbol is non-negative?
   (I know this is unlikely to have any impact on the plans.)

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/AssignUniqueIdStatsRule.java:
```diff
@@ -0,0 +1,50 @@
+import java.util.Map;
+import java.util.Optional;
+
+public class AssignUniqueIdStatsRule
```
Would it be possible to have an integration tests for this? e.g. `TestTpchLocalStats`?
(you may want to rebase onto https://github.com/prestodb/presto/pull/10659 to avoid conflicts)

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java:
```diff
@@ -241,15 +242,10 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
             if (rightStats.getOutputRowCount() > 0) {
                 // all left side rows are matched
+                // TODO: account non-equi conditions here somehow
                 return leftStats.mapOutputRowCount(rowCount -> 0.0);
```
This was correct based on the assumption that `! filter.isPresent()`, which is not necessarily true anymore.

kokosing replied:
It looks here should be some if with `! filter.isPresent()` or logic from line 322 used.

Can you add coverage this case where there is no equi criteria, but filter present?

findepi commented on presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java:
```diff
@@ -241,15 +242,10 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
+                // TODO: account non-equi conditions here somehow
                 return leftStats.mapOutputRowCount(rowCount -> 0.0);
             }
             if (rightStats.getOutputRowCount() == 0) {
```
This (`if right is empty then outer == left side`) is universally true, should be pulled up, before `if (criteria.isEmpty())` condition.

kokosing commented on presto-tests/src/test/java/com/facebook/presto/tests/TestTpchLocalStats.java:
```diff
@@ -228,6 +228,16 @@ public void testLeftJoinStats()
                         .verifyColumnStatistics("l_orderkey", absoluteError(6.0)));
+
+        // simple non-equi join
+        statisticsAssertion.check("SELECT * FROM partsupp left join lineitem on ps_partkey = l_partkey AND ps_suppkey < l_suppkey",
```
make SQL keywords (LEFT, JOIN, ON) to be uppercase

kokosing commented on presto-main/src/main/java/com/facebook/presto/cost/JoinStatsRule.java:
```diff
@@ -260,14 +256,18 @@ PlanNodeStatsEstimate calculateJoinComplementStats(
                     EquiJoinClause drivingClause = criteria.get(drivingClauseId);
-                    List<EquiJoinClause> remainingClauses = copyWithout(criteria, drivingClauseId);
-                    return calculateJoinComplementStats(leftStats, rightStats, drivingClause, remainingClauses);
+                    return calculateJoinComplementStats(leftStats, rightStats, drivingClause, criteria.size() - 1 + numberOfFilterClauses);
```
maybe instead of `-1`, use `remainingCluases.size()`?

sopel39 replied:
I didn't want to pass `remainingCluases` if I'm just using clauses list for getting size

sopel39 commented:
@rschlussel2 I will run benchmarks to see if there are significant perf differences

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/cost/AssignUniqueIdStatsRule.java:
```diff
@@ -0,0 +1,53 @@
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+
+public class AssignUniqueIdStatsRule
+        implements ComposableStatsCalculator.Rule
```
@sopel39 Why `ComposableStatsCalculator.Rule` is not a generic with `PlanNode` parameter?

```
public class AssignUniqueIdStatsRule 
                 implements ComposableStatsCalculator.Rule<AssignUniqueId>`
```

would be better. Then `(AssignUniqueId) node` cast in `apply` won't be needed.

findepi replied:
It could be, just like in `com.facebook.presto.sql.planner.iterative.Rule`. @mbasmanova do you want to fix this?

mbasmanova replied:
@findepi There are not too many places to update yet. Here you go: #10748

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/cost/AssignUniqueIdStatsRule.java:
```diff
@@ -0,0 +1,53 @@
+public class AssignUniqueIdStatsRule
+        implements ComposableStatsCalculator.Rule
+{
+    private static final Pattern<AssignUniqueId> PATTERN = Pattern.typeOf(AssignUniqueId.class);
```
@sopel39 Let's add `Patterns#assignUniqueId()` method.

```
public class Patterns
{
...
    public static Pattern<AssignUniqueId> assignUniqueId()
    {
        return typeOf(AssignUniqueId.class);
    }

```

mbasmanova commented on presto-main/src/main/java/com/facebook/presto/cost/CostCalculatorUsingExchanges.java:
```diff
@@ -119,6 +121,12 @@ public PlanNodeCostEstimate visitGroupReference(GroupReference node, Void contex
+        @Override
+        public PlanNodeCostEstimate visitAssignUniqueId(AssignUniqueId node, Void context)
+        {
+            return cpuCost(getStats(node).getOutputRowCount() * BigintType.BIGINT.getFixedSize());
```
Why are you hardcoding `BigintType.BIGINT.getFixedSize()` instead of `getStats(node).getSymbolStatistics(node.getIdColumn()).getAverageRowSize()`?

findepi replied:
AFAIK we don't fill `#getAverageRowSize()` based on type yet

mbasmanova replied:
@findepi It looks like we do. `AssignUniqueIdStatsRule` contains `.setAverageRowSize(BIGINT.getFixedSize())`.

findepi replied:
Right. I believe my comment remains generally applicable, although in this case we could defer to average row size from stats.

sopel39 commented:
Addressed all comments, ready for final review and landing
TPCH/TPCDS results with CBO for queries where plan changed:
```
TPCH/q21
Before(1000): Query 20180615_105041_00023_5sbuz failed: Query exceeded local memory limit
After (1000): 1:40

Before (300): 26
After (300): 23 (~1.1x impr)

TPCDS/q16
Before (3000): 51
after (3000): 32 (1.6x impr)

TPCDS/q94
Before (3000): 19
After (3000): 11 (1.7x impv)
```

sopel39 commented:
Could you stamp it @fiedukow @rschlussel2?

kokosing commented:
On what branch did you run the benchmark?

sopel39 commented:
> On what branch did you run the benchmark?
195e
